### ISSUE TITLE
Access CommDiagnostics capabilities through objects?

### ISSUE THREAD

**Issue by bradcray** [chapel-lang/chapel#16956]
This week, we discussed whether the CommDiagnostics module would benefit from becoming
more explicitly object-oriented rather than relying on implicit, hidden state.  This was generally
received positively, where we were able to imagine a unified interface as proposed in issue #16955
being supported on this object such that you'd write something like:

```chapel
commDiags.start(...);
commDiags.stop(...);
```

[where here, and throughout this issue, consider `commDiags` to be a placeholder name for
such an object, not the actual proposed name].

There was some question about whether the object on which these routines acted should be:

* a global singleton object
* a singleton object per locale
* an object type that the user could create arbitrary instances of, like `timer`/`stopwatch`

Commentary on each of these options:

* **using global singleton object:** This raised a question about how a global singleton object differs from having standalone functions defined on a module (where I think there was an answer, but I didn't capture it... TODO: we should fill it in here)

* **singleton object per locale:** It was proposed that this might support the ability to write patterns like:
  - `here.commDiags.start(...)`
  - `Locales.commDiags.start(...)`
  - `Locales[mySubset].commDiags.start(...)`
  - `myLocaleArr.commDiags.start(...)`
  effectively moving the specification of which locales the measurements should be taken on from the argument list to the 
  receiver and relying on promotion.  There were definitely concerns about this approach as well.

* **arbitrary object**: While there was some interest in making this object's interface more like `timer`/`stopwatch`, there were also concerns that measuring time passing seemed more independent of who which task was doing the measuring and what else was going on; whereas it might seem surprising to turn on a communication counter and get counts for communications done by your locale (and/or others) which weren't due to the code you had bracketed with the start/stop calls (?).

**Comment by e-kayrakli** [comment]
> This raised a question about how a global singleton object differs from having standalone functions defined on a module

One argument for a global object is to avoid `use` vs `import`-based naming discussion. i.e. the only way we can have concise function names like `start` and `stop` is by having them as methods on some object and if the names are more verbose, it wouldn't play along well with `import`-based implementations.

**Comment by lydia-duncan** [comment]
> if the names are more verbose, it wouldn't play along well with import-based implementations.

To be fair, users can rename these functions to a shorter name as part of the import or use.  But I agree that we don't want the names to be absurdly long

**Comment by dlongnecke-cray** [comment]
I just want to observe that a huge benefit of going this direction is that we can use comm counter objects with context managers.

**Comment by bradcray** [comment]
Not objecting to your comment or this issue David, but even without going this direction, couldn't we use comm counters with context managers by creating a little object that calls the global / non-OOP routines or something like that?

**Comment by dlongnecke-cray** [comment]
Yep, that's true. In my head it seemed more natural if it was on the object itself, but we already create little "context objects" for things like `map.guard()` (a method I demoed awhile back as a proof of concept).

**Comment by bradcray** [comment]
Makes sense, and I agree it would be more natural.  I just wanted to make sure I was understanding.
